libr/bin/dwarf.c

```diff
@@ -1319,321 +1519,523 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
 		break;
 	case DW_FORM_string:
-		value->encoding.str_struct.string = *buf? strdup ((const char*)buf) : NULL;
-		buf += (strlen ((const char*)buf) + 1);
+		value->string.content = *buf ? strdup ((const char *)buf) : NULL;
+		buf += (strlen ((const char *)buf) + 1);
 		break;
-	case DW_FORM_block:
-		buf = r_uleb128 (buf, buf_end - buf, &value->encoding.block.length);
-		if (!buf || buf >= buf_end) {
+	case DW_FORM_block1:
+		value->block.length = READ8 (buf);
+		value->block.data = calloc (sizeof (ut8), value->block.length + 1);
+		if (!value->block.data) {
 			return NULL;
 		}
-		value->encoding.block.data = calloc (sizeof (ut8), value->encoding.block.length);
-		if (value->encoding.block.data) {
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
+		if (value->block.data) {
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
 			}
 		}
 		break;
-	case DW_FORM_block1:
-		value->encoding.block.length = READ (buf, ut8);
-		value->encoding.block.data = calloc (sizeof (ut8), value->encoding.block.length + 1);
-		if (value->encoding.block.data) {
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
+	case DW_FORM_block2:
+		value->block.length = READ16 (buf);
+		if (value->block.length > 0) {
+			value->block.data = calloc (sizeof (ut8), value->block.length);
+			if (!value->block.data) {
+				return NULL;
+			}
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
 			}
 		}
 		break;
-	case DW_FORM_flag:
-		value->encoding.flag = READ (buf, ut8);
+	case DW_FORM_block4:
+		value->block.length = READ32 (buf);
+		if (value->block.length > 0) {
+			ut8 *data = calloc (sizeof (ut8), value->block.length);
+			if (!data) {
+				return NULL;
+			}
+			for (j = 0; j < value->block.length; j++) {
+				data[j] = READ (buf, ut8);
+			}
+			value->block.data = data;
+		}
 		break;
-	case DW_FORM_sdata:
-		buf = r_leb128 (buf, buf_end - buf, &value->encoding.sdata);
+	case DW_FORM_block: // variable length ULEB128
+		buf = r_uleb128 (buf, buf_end - buf, &value->block.length);
+		if (!buf || buf >= buf_end) {
+			return NULL;
+		}
+		value->block.data = calloc (sizeof (ut8), value->block.length);
+		if (value->block.data) {
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
+			}
+		}
 		break;
+	case DW_FORM_flag:
+		value->flag = READ (buf, ut8);
+		break;
+	// offset in .debug_str
 	case DW_FORM_strp:
-		value->encoding.str_struct.offset = READ32 (buf);
-		if (debug_str && value->encoding.str_struct.offset < debug_str_len) {
-			value->encoding.str_struct.string = strdup (
-				(const char *)(debug_str +
-					value->encoding.str_struct.offset));
+		value->string.offset = dwarf_read_piece(hdr->is_64bit, &buf, buf_end);
+		if (debug_str && value->string.offset < debug_str_len) {
+			value->string.content =
```

Comment:
no check that the string is actually terminated? At least terminated before end of debug_str section.